Cannot build on OSX with Boost 1.67

**Update:** 
I was able to overcome this issue by uninstalling boost 1.67 and installing Boost 1.55 via `brew install boost@1.55` followed by `brew link boost --force`.  Something should be updated so others can avoid same issue, either the install doc or update for latest Boost compatibility.

**OP:**
I am getting an error when trying to build on OSX with the latest boost via `brew install boost` per the `doc/build-osx.md` file without success.  This seems to be similar to this other issue related to Boost version >= 1.66: [bitcoin/bitcoin/pull/11847](https://github.com/bitcoin/bitcoin/pull/11847)

**Errors from Boost 1.67 on mac osx High Sierra:**
```
  CXX      libbitcoin_server_a-txmempool.o
In file included from txmempool.cpp:6:
In file included from ./txmempool.h:19:
In file included from /usr/local/include/boost/multi_index/ordered_index.hpp:17:
/usr/local/include/boost/multi_index/detail/ord_index_impl.hpp:1194:10: error: no matching function for call to object of type 'const
      boost::multi_index::detail::ordered_index_impl<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, boost::multi_index::detail::nth_layer<2,
      CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_non_unique_tag,
      boost::multi_index::detail::null_augment_policy>::key_compare' (aka 'const CompareTxMemPoolEntryByFee')
      if(comp_(key(v),key(y->value())))return false;
         ^~~~~
/usr/local/include/boost/multi_index/detail/ord_index_impl.hpp:838:9: note: in instantiation of member function
      'boost::multi_index::detail::ordered_index_impl<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, boost::multi_index::detail::nth_layer<2,
      CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_non_unique_tag,
      boost::multi_index::detail::null_augment_policy>::in_place' requested here
      b=in_place(x->value(),x,Category());
        ^
/usr/local/include/boost/multi_index/detail/ord_index_impl.hpp:873:18: note: in instantiation of member function
      'boost::multi_index::detail::ordered_index_impl<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, boost::multi_index::detail::nth_layer<2,
      CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_non_unique_tag,
      boost::multi_index::detail::null_augment_policy>::modify_' requested here
      if(!super::modify_(x)){
                 ^
/usr/local/include/boost/multi_index_container.hpp:845:18: note: in instantiation of member function 'boost::multi_index::detail::ordered_index_impl<mempoolentry_txid,
      std::__1::less<uint256>, boost::multi_index::detail::nth_layer<1, CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid,
      mpl_::na, mpl_::na>, boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_unique_tag,
      boost::multi_index::detail::null_augment_policy>::modify_' requested here
      if(!super::modify_(x)){
                 ^
/usr/local/include/boost/multi_index/detail/index_base.hpp:280:21: note: in instantiation of function template specialization
      'boost::multi_index::multi_index_container<CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >::modify_<update_descendant_state>' requested here
    {return final().modify_(mod,x);}
                    ^
/usr/local/include/boost/multi_index/detail/ord_index_impl.hpp:431:18: note: in instantiation of function template specialization
      'boost::multi_index::detail::index_base<CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >::final_modify_<update_descendant_state>' requested here
    return this->final_modify_(
                 ^
txmempool.cpp:108:11: note: in instantiation of function template specialization 'boost::multi_index::detail::ordered_index_impl<mempoolentry_txid, std::__1::less<uint256>,
      boost::multi_index::detail::nth_layer<1, CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_unique_tag,
      boost::multi_index::detail::null_augment_policy>::modify<update_descendant_state>' requested here
    mapTx.modify(updateIt, update_descendant_state(modifySize, modifyFee, modifyCount));
          ^
./txmempool.h:147:10: note: candidate function not viable: 'this' argument has type 'const
      boost::multi_index::detail::ordered_index_impl<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, boost::multi_index::detail::nth_layer<2,
      CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_non_unique_tag,
      boost::multi_index::detail::null_augment_policy>::key_compare' (aka 'const CompareTxMemPoolEntryByFee'), but method is not marked const
    bool operator()(const CTxMemPoolEntry& a, const CTxMemPoolEntry& b)
         ^
In file included from txmempool.cpp:6:
In file included from ./txmempool.h:19:
In file included from /usr/local/include/boost/multi_index/ordered_index.hpp:17:
/usr/local/include/boost/multi_index/detail/ord_index_impl.hpp:1199:26: error: no matching function for call to object of type 'const
      boost::multi_index::detail::ordered_index_impl<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, boost::multi_index::detail::nth_layer<2,
      CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_non_unique_tag,
      boost::multi_index::detail::null_augment_policy>::key_compare' (aka 'const CompareTxMemPoolEntryByFee')
    return y==header()||!comp_(key(y->value()),key(v));
                         ^~~~~
./txmempool.h:147:10: note: candidate function not viable: 'this' argument has type 'const
      boost::multi_index::detail::ordered_index_impl<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, boost::multi_index::detail::nth_layer<2,
      CTxMemPoolEntry, boost::multi_index::indexed_by<boost::multi_index::ordered_unique<mempoolentry_txid, mpl_::na, mpl_::na>,
      boost::multi_index::ordered_non_unique<boost::multi_index::identity<CTxMemPoolEntry>, CompareTxMemPoolEntryByFee, mpl_::na>, mpl_::na, mpl_::na, mpl_::na, mpl_::na,
      mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na, mpl_::na>,
      std::__1::allocator<CTxMemPoolEntry> >, boost::mpl::vector0<mpl_::na>, boost::multi_index::detail::ordered_non_unique_tag,
      boost::multi_index::detail::null_augment_policy>::key_compare' (aka 'const CompareTxMemPoolEntryByFee'), but method is not marked const
    bool operator()(const CTxMemPoolEntry& a, const CTxMemPoolEntry& b)
         ^
2 errors generated.
make[2]: *** [libbitcoin_server_a-txmempool.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1

```

--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -162,7 +162,7 @@ struct update_descendant_state
     {
     }
 
-    void operator()(CTxMemPoolEntry &e) { e.UpdateState(modifySize, modifyFee, modifyCount); }
+    void operator()(CTxMemPoolEntry &e) const { e.UpdateState(modifySize, modifyFee, modifyCount); }
 private:
     int64_t modifySize;
     CAmount modifyFee;
@@ -171,13 +171,13 @@ private:
 
 struct set_dirty
 {
-    void operator()(CTxMemPoolEntry &e) { e.SetDirty(); }
+    void operator()(CTxMemPoolEntry &e) const { e.SetDirty(); }
 };
 
 struct update_fee_delta
 {
     update_fee_delta(int64_t _feeDelta) : feeDelta(_feeDelta) {}
-    void operator()(CTxMemPoolEntry &e) { e.UpdateFeeDelta(feeDelta); }
+    void operator()(CTxMemPoolEntry &e) const { e.UpdateFeeDelta(feeDelta); }
 private:
     int64_t feeDelta;
 };
@@ -185,7 +185,7 @@ private:
 struct update_lock_points
 {
     update_lock_points(const LockPoints &_lp) : lp(_lp) {}
-    void operator()(CTxMemPoolEntry &e) { e.UpdateLockPoints(lp); }
+    void operator()(CTxMemPoolEntry &e) const { e.UpdateLockPoints(lp); }
 private:
     const LockPoints &lp;
 };
@@ -204,7 +204,7 @@ struct mempoolentry_txid
 class CompareTxMemPoolEntryByDescendantScore
 {
 public:
-    bool operator()(const CTxMemPoolEntry &a, const CTxMemPoolEntry &b)
+    bool operator()(const CTxMemPoolEntry &a, const CTxMemPoolEntry &b) const
     {
         bool fUseADescendants = UseDescendantScore(a);
         bool fUseBDescendants = UseDescendantScore(b);
@@ -227,7 +227,7 @@ public:
     }
 
     // Calculate which score to use for an entry (avoiding division).
-    bool UseDescendantScore(const CTxMemPoolEntry &a)
+    bool UseDescendantScore(const CTxMemPoolEntry &a) const
     {
         double f1 = (double)a.GetModifiedFee() * a.GetSizeWithDescendants();
         double f2 = (double)a.GetModFeesWithDescendants() * a.GetTxSize();
@@ -242,7 +242,7 @@ public:
 class CompareTxMemPoolEntryByScore
 {
 public:
-    bool operator()(const CTxMemPoolEntry &a, const CTxMemPoolEntry &b)
+    bool operator()(const CTxMemPoolEntry &a, const CTxMemPoolEntry &b) const
     {
         double f1 = (double)a.GetModifiedFee() * b.GetTxSize();
         double f2 = (double)b.GetModifiedFee() * a.GetTxSize();
@@ -257,7 +257,7 @@ public:
 class CompareTxMemPoolEntryByEntryTime
 {
 public:
-    bool operator()(const CTxMemPoolEntry &a, const CTxMemPoolEntry &b) { return a.GetTime() < b.GetTime(); }
+    bool operator()(const CTxMemPoolEntry &a, const CTxMemPoolEntry &b) const { return a.GetTime() < b.GetTime(); }
 };
 
 // Multi_index tag names

